drivers/input/input_mcux_kpp.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2025, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <fsl_kpp.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/dt-bindings/clock/imx_ccm.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(kpp, CONFIG_INPUT_LOG_LEVEL);
+
+#define DT_DRV_COMPAT nxp_mcux_kpp
+
+#define INPUT_KPP_COLUMNNUM_MAX  KPP_KEYPAD_COLUMNNUM_MAX
+#define INPUT_KPP_ROWNUM_MAX     KPP_KEYPAD_ROWNUM_MAX
+
+struct kpp_config {
+	KPP_Type *base;
+	const struct device *ccm_dev;
+	clock_control_subsys_t clk_sub_sys;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct kpp_data {
+	uint32_t clock_rate;
+};
+
+static void get_source_clk_rate(const struct device *dev, uint32_t *clk_rate)
+{
+	const struct kpp_config *dev_cfg = dev->config;
+	const struct device *ccm_dev = dev_cfg->ccm_dev;
+	clock_control_subsys_t clk_sub_sys = dev_cfg->clk_sub_sys;
+
+	if (!device_is_ready(ccm_dev)) {
+		LOG_ERR("CCM driver is not installed");
+		*clk_rate = 0;
+		return;
+	}
+
+	clock_control_get_rate(ccm_dev, clk_sub_sys, clk_rate);
+}
+
+static void input_kpp_data_analyze(const struct device *dev, uint8_t *data)
+{
+	const struct kpp_config *config = dev->config;
+	int row = -1;
+	int col = -1;
+	int row_mask;
+
+	for (int i = 0; i < INPUT_KPP_COLUMNNUM_MAX; i++) {
+		if (data[i] != 0) {
+			row_mask = data[i];
+			for (int j = 0; j < INPUT_KPP_ROWNUM_MAX; j++) {
+				if ((row_mask & BIT(j)) == 0) {
+					row = j;
+					col = i;
+					continue;
+				}
+			}
+		}
+	}
+
+	if (row == -1 || col == -1) {
+		LOG_DBG("No key signal detected in matrix scan");
+	} else {
+		input_report_abs(dev, INPUT_ABS_X, row, false, K_FOREVER);
+		input_report_abs(dev, INPUT_ABS_Y, col, false, K_FOREVER);
+		KPP_SetSynchronizeChain(config->base, kKPP_ClearKeyDepressSyncChain);
+		input_report_key(dev, INPUT_BTN_TOUCH, 1, true, K_FOREVER);
+	}
+}
+
+static void kpp_isr(const struct device *dev)
+{
+	k_busy_wait(400);
```

Comment:
This is not great, 400us spinning in the interrupt does not strike me as a great idea, plus there's quite some code in here, would it make sense to have a deferred work (or a delayable work so you can have the delay there) and do this outside of interrupt context?

sorry only noticed it now